cloudera/datadog_checks/cloudera/api_client_v7.py

```diff
@@ -0,0 +1,131 @@
+from datadog_checks.cloudera.entity_status import ENTITY_STATUS
+from datadog_checks.cloudera.metrics import NATIVE_METRICS, TIMESERIES_METRICS
+
+from .common import CLUSTER_HEALTH, HOST_HEALTH
+
+
+class ApiClientV7(ApiClient):
+    def __init__(self, check, api_client):
+        super(ApiClientV7, self).__init__(check, api_client)
+
+    def collect_data(self):
+        clusters_resource_api = cm_client.ClustersResourceApi(self._api_client)
+        read_clusters_response = clusters_resource_api.read_clusters(cluster_type='any', view='full')
+        self._log.debug("Cloudera full clusters response:\n%s", read_clusters_response)
+
+        # Use len(read_clusters_response.items) * 3 workers since
+        # for each cluster, we are executing 3 tasks in parallel.
+        with ThreadPoolExecutor(max_workers=len(read_clusters_response.items) * 3) as executor:
+            for cluster in read_clusters_response.items:
+                cluster_name = cluster.name
+
+                tags = self._collect_cluster_tags(cluster, self._check.config.tags)
+
+                executor.submit(self._collect_cluster_metrics, cluster_name, tags)
+                executor.submit(self._collect_cluster_service_check, cluster, tags)
+                executor.submit(self._collect_hosts, cluster_name)
+
+    @staticmethod
+    def _collect_cluster_tags(cluster, custom_tags):
+        cluster_tags = [f"{cluster_tag.name}:{cluster_tag.value}" for cluster_tag in cluster.tags]
+
+        cluster_tags.extend(custom_tags)
+
+        return cluster_tags
+
+    def _collect_cluster_service_check(self, cluster, tags):
+        cluster_entity_status = ENTITY_STATUS[cluster.entity_status]
+        message = cluster.entity_status if cluster_entity_status != AgentCheck.OK else None
+        self._check.service_check(
+            CLUSTER_HEALTH, cluster_entity_status, tags=[f'cloudera_cluster:{cluster.name}', *tags], message=message
+        )
+
+    def _collect_cluster_metrics(self, cluster_name, tags):
+        metric_names = ','.join(f'last({metric}) AS {metric}' for metric in TIMESERIES_METRICS['cluster'])
+        query = f'SELECT {metric_names} WHERE clusterName="{cluster_name}" AND category=CLUSTER'
+        self._query_time_series(query, tags=tags)
+
+    def _collect_hosts(self, cluster_name):
+        clusters_resource_api = cm_client.ClustersResourceApi(self._api_client)
+        list_hosts_response = clusters_resource_api.list_hosts(cluster_name, view='full')
+        self._log.debug("Cloudera full hosts response:\n%s", list_hosts_response)
+
+        # Use len(list_hosts_response.items) * 4 workers since
+        # for each host, we are executing 4 tasks in parallel.
+        with ThreadPoolExecutor(max_workers=len(list_hosts_response.items) * 4) as executor:
+            for host in list_hosts_response.items:
+                tags = self._collect_host_tags(host, self._check.config.tags)
+                executor.submit(self._collect_host_metrics, host, tags)
```

Comment:
This also spawns threads; should we just move that logic here?

Reply:
Yes `_collect_host_metrics()` does also spawn its own threads for native and timeseries metrics, although I think keeping it within `_collect_host_metrics()` instead of moving that logic here keeps the abstraction easier to understand.